### PR TITLE
start typing dust API

### PR DESCRIPTION
**Pull request by fontanierh**

### CODE REVIEW

File: front/lib/dust_api.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import { Project } from "@app/types/project";
+
+const { DUST_API: DUST_API_URL } = process.env;
+
+type ErrorRsponse = { error: string };
```
Comment (spolu) [inline front/lib/dust_api.ts:5]: We return error objects not strings from core?

Comment (fontanierh) [reply front/lib/dust_api.ts:5]: Ah my bad

File: front/pages/api/apps/[user]/[sId]/datasets/[name]/[hash]/index.ts [hunk 2]
```diff
@@ -1,22 +1,52 @@
...
+    name: { type: "string" },
+  },
+  required: ["hash", "user", "sId", "name"],
+};
+
 async function handler(
   req: NextApiRequest,
   res: NextApiResponse<GetDatasetByHashResponseBody>
 ): Promise<void> {
+  const getDatasetQueryRes = parse_payload(getDatasetQuerySchema, req.query);
```
Comment (spolu) [inline front/pages/api/apps/[user]/[sId]/datasets/[name]/[hash]/index.ts:36]: let's call it queryRes and query to be a bit less verbose?

Comment (fontanierh) [reply front/pages/api/apps/[user]/[sId]/datasets/[name]/[hash]/index.ts:36]: i copied from search for consistency, I can change it over there as well

File: front/lib/dust_api.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  async createProject(): Promise<DustAPIResponse<{ project: Project }>> {
+    const response = await fetch(`${DUST_API_URL}/projects`, {
+      method: "POST",
+    });
+    return response.json() as Promise<DustAPIResponse<{ project: Project }>>;
+  },
+
+  async getDatasets(
+    projectId: string
+  ): Promise<DustAPIResponse<GetDatasetsResponse>> {
```
Comment (spolu) [inline front/lib/dust_api.ts:40]: We should use @lasryaric 's Error/Ok framework here?

Comment (fontanierh) [reply front/lib/dust_api.ts:45]: @lasryaric @spolu  can you check https://github.com/dust-tt/dust/pull/229/commits/643a36879ff1ab06019d669f22cbc275c73cbe7a ? It updates the `Result` framework to use type guards (the compiler forces you at compile time to check if there's an error before you can use `value`, slightly safer than having an exception at runtime)

File: front/lib/dust_api.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+): response is ErrorResponse {
+  return "error" in response;
+}
+
+type GetDatasetsResponse = {
+  datasets: { [key: string]: { hash: string; created: number }[] };
+};
+
+type GetDatasetResponse = {
+  dataset: {
```
Comment (spolu) [inline front/lib/dust_api.ts:26]: Shouldn't we use `DatasetType` (extending it if needed)?

Comment (fontanierh) [reply front/lib/dust_api.ts:20]: As far as I can tell, the types are very different and only have `data` in common. Am I missing something ?

File: front/lib/dust_api.ts (hunk 3 above)
Comment (fontanierh) [reply front/lib/dust_api.ts:39]: Yes done in the commit right after

File: front/lib/dust_api.ts (hunk 4 above)
Comment (spolu) [reply front/lib/dust_api.ts:20]: Ah sorry you're right. types/dataset.rs is associated with the `front` SQL object `Dataset` and is "hydrated" with the data coming from the DustAPI :+1: 

You're perfectly right.

Maybe we want named types in this file as well. DustAPIDataset DustAPIRun DustAPIDataSource etc... ?

That being said, for now this one is used only once so I'm fine without DRY-ing up these types :+1:

Comment (fontanierh) [reply front/lib/dust_api.ts:20]: > Maybe we want named types in this file as well. DustAPIDataset DustAPIRun DustAPIDataSource etc... ?

That sounds good to me

**Review by spolu — CHANGES_REQUESTED** [review]
LGTM, let's just remove the `.response` and have `p.value.something`

File: front/lib/dust_api.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+          "Content-Type": "application/json",
+        },
+      }
+    );
+    return _resultFromResponse(response);
+  },
+};
+
+async function _resultFromResponse<T>(
+  response: Response
```
[the remark this answers is not in the document]
Comment (fontanierh) [reply front/lib/dust_api.ts:78]: It ships with TS

**Comment by fontanierh** [comment]
Fixed @spolu

**Review by spolu — APPROVED** [review]
LGTM!

Sorry for the delay, missed the last notification.

**Comment by fontanierh** [comment]
No worries !

**Pull request merged**